Reformular o capítulo de programação orientada a objetos

luizffgv opened the pull request.

luizffgv committed:
Fix introdução POO

Fix #1

luizffgv committed:
Começar a reformular o cap. 5

tarcisiofischer commented on src/cap5/objetos.md:
```diff
@@ -0,0 +1,32 @@
+# Objetos
+
+**Não se confunda. Este capítulo utiliza o termo _objeto_ no contexto geral de
+programação orientada a objetos. Em C++ esse termo possui outro significado, mas
```
Acho que seria bom aproveitar pra comentar minimamente o que é o outro significado pra não deixar o leitor "na mão"

luizffgv replied:
Quebrei bastante a cabeça e é absurdamente difícil definir objeto no C++ moderno. No C++14 o padrão diz "An object is a
region of storage", mas no C++17 pra frente esse trecho foi removido.

Consigo adicionar uma footnote com o seguinte trecho no final da página:
>  Objetos são parte fundamental do modelo de funcionamento do C++. Em C++ todas as variáveis que não são referências são objetos, mas objetos não se limitam apenas a isso.

Mas não sei se essa definição é satisfatória. O que acha?

tarcisiofischer replied:
Sendo sincero, eu nem comentaria nada disso de "Não se confunda. Este capítulo utiliza o termo _objeto_ no contexto geral de programação orientada a objetos. Em C++ esse termo possui outro significado.", pq acho que no contexto do capítulo, eu acho claro que se está falando de objeto no contexto de orientação à objetos.

Mas se discordares disso, tudo bem. Minha sugestão seria colocar o que tu quer dizer exatamente com tuas preocupações, e se alguém no futuro tiver algo melhor pode adicionar. Tentei juntar tuas palavras pra criar uma sugestão de texto:

> **Este capítulo utiliza o termo _objeto_ no contexto geral de programação orientada a objetos. Note que o termo, no padrão do C++, possui um significado diferente. No C++14 o padrão diz que "An object is a region of storage", mas no C++17 pra frente esse trecho foi removido do padrão.

E, se quiser realmente ser preciosista, daria até pra adicionar algo como

> Além disso, a palavra "objeto" é também utilizada em outro contexto para designar os arquivos gerados no processo de compilação (com extensão .obj). Apesar de ser a mesma palavra, não tem nada a ver com os objetos explicados nesse capítulo.

Eu sou contra usar qualquer coisa produzida pelo ChatGPT atualmente pq acho que perde o propósito colaborativo de criação e do processo criativo envolvido no desenvolvimento do livro. Se o leitor quer a opinião do ChatGPT, ele que use o bot :)

luizffgv replied:
Tudo o que disse é válido. Quando essa PR estiver quase pronta vou analisar novamente se realmente há algum ponto de confusão pra esse aviso sobre objetos ser necessário, e se sim, garantir que ele esteja de acordo com o que quer no livro.

Respeito sua posição quanto ao GPT, porém eu diria que a definição que adicionei no commit [ecd0254](https://github.com/tarcisiofischer/livrocpp/pull/18/commits/ecd0254a98ce409f2d0a381c6dbd4b67b67bb455) utilizou o modelo como fonte e não foram utilizadas as mesmas palavras (e nem mesmo o mesmo significado), então não acredito que o GPT seria considerado um coautor. O modelo apenas me forneceu uma base utilizando as informações e correções que eu seletivamente forneci, e isso resulta numa definição distante das que você encontraria na Internet.

Se mesmo com isso em mente sua posição for a mesma, vou obviamente segui-la.

luizffgv committed:
Introduzir classes (conceito)

luizffgv committed:
Add definição de "objeto"

luizffgv committed:
Add Relações (composição, herança)

luizffgv committed:
Fix formatação em lista

Estava causando inconsistências no HTML produzido

tarcisiofischer commented on src/cap5/objetos.md:
```diff
@@ -0,0 +1,57 @@
+
+Na programação orientada a objetos (POO), funções e dados podem ser unidos para
+formar o que chamamos objetos. Comumente, essas funções são chamadas _métodos_,
+e esses dados são chamados _campos_. Enquanto na programação procedural uma
```
Eu acho que nunca ouvi a nomenclatura `campos`... Seriam `atributos`?

tarcisiofischer commented on src/cap5/objetos.md:
```diff
@@ -0,0 +1,57 @@
+Na programação orientada a objetos (POO), funções e dados podem ser unidos para
+formar o que chamamos objetos. Comumente, essas funções são chamadas _métodos_,
+e esses dados são chamados _campos_. Enquanto na programação procedural uma
+função normalmente tem acesso apenas aos seus argumentos e ao estado global do
```
Sugestão: `Uma função possui acesso à argumentos, variáveis globais e variáveis locais criadas no escopo da função`

luizffgv replied:
Adicionei uma tabela e deixei um pouco mais próximo de sua sugestão

tarcisiofischer commented on src/cap5/objetos.md:
```diff
@@ -0,0 +1,57 @@
+formar o que chamamos objetos. Comumente, essas funções são chamadas _métodos_,
+e esses dados são chamados _campos_. Enquanto na programação procedural uma
+função normalmente tem acesso apenas aos seus argumentos e ao estado global do
+programa, em POO um método tem, adicionalmente, acesso ao objeto ao qual
```
`acesso ao objeto ao qual pertence e seus atributos`

tarcisiofischer commented on src/cap5/objetos.md:
```diff
@@ -0,0 +1,57 @@
+programa, em POO um método tem, adicionalmente, acesso ao objeto ao qual
+pertence.
+
+Como exemplo, imagine uma função _depositar_ que deposita dinheiro em um cofre.
```
Favor adicionar o código relacionado para facilitar a leitura.

luizffgv replied:
Adicionei mais abaixo. Não estou muito satisfeito com o quão longe o código ficou lá embaixo, então vou tentar melhorar depois.

tarcisiofischer commented on src/cap5/objetos.md:
```diff
@@ -0,0 +1,57 @@
+Como exemplo, imagine uma função _depositar_ que deposita dinheiro em um cofre.
+Para funcionar, ela precisa saber quanto dinheiro e em qual cofre depositar.
+Essas informações podem ser enviadas como dois parâmetros: cofre e quantia. Para
+depositar R$ 10 no cofre _c_, precisamos chamar a função e fornecer os
```
Sugiro evitar `c`, e usar `cofre` por dois motivos:
1- Facilita a leitura (na minha opinião)
2- Usar nome de variável com uma só letra em código em geral é uma má prática (discutível)

luizffgv replied:
Utilizei `cof` pois creio ser descritivo o suficiente e não há a possibilidade de confundir com o parâmetro

tarcisiofischer commented on src/cap5/objetos.md:
```diff
@@ -0,0 +1,57 @@
+argumentos _c_ e 10. Na linguagem C, isso seria `depositar(c, 10)`. Isso é um
+exemplo de programação procedural.
+
+Em POO, _c_ seria um objeto contendo o método _depositar_ e o campo _v_ (que
```
Novamente, sugiro adicionar código inteiro para mostrar a sintaxe.

luizffgv replied:
https://github.com/tarcisiofischer/livrocpp/pull/18#discussion_r1059240139

tarcisiofischer commented on src/cap5/objetos.md:
```diff
@@ -0,0 +1,57 @@
+
+---
+
+Na programação orientada a objetos (POO), funções e dados podem ser unidos para
```
Lendo novamente, eu acho que ficou um pouco confuso. Onde fica a definição do que é uma classe, e a diferença entre classe e objeto? A versão antiga me parecia mais completa:

"No contexto de C++ objetos são instâncias de classes, classes são abstrações que contém a descrição de quais atributos e métodos um objeto possui e atributos são variáveis internas de uma classe."

tarcisiofischer commented on src/cap5/objetos.md:
```diff
@@ -0,0 +1,57 @@
+
+## Como saber quais campos e métodos um objeto possui?
+
+Em C++ utiliza-se o estilo de POO baseado em classes. Nesse estilo, objetos são
```
Seria útil mostrar a sintaxe de como representar uma classe em C++; Como instanciar um objeto daquela classe; Como invocar um método e acessar os membros da classe. (Estou repetindo alguns comentários já colocados, mas acho que multiplos exemplos é útil para o leitor iniciante).

A versão antiga tinha a vantagem que possuia mais exemplos, sinto falta deles no texto

luizffgv replied:
Botei um exemplo que acessa dois métodos. Vou ver se consigo botar um acesso de membro nessa página ou na _Especificadores de acesso_

tarcisiofischer commented on src/cap5/relacoes.md:
```diff
@@ -0,0 +1,72 @@
+_Livro_ possui uma instância de _Capa_ e uma de _Contracapa_, ou, mais
+diretamente, um _Livro_ possui uma _Capa_ e _Contracapa_.
+
+![UML - Composição](images/composicao.svg)
```
Eu acho interessante ter exemplos com UML, mas lembre-se que isso é um livro de C++. Dito isso, mais importante que mostrar essas imagens é mostrar um exemplo (em código C++ compilável e executável) para que o leitor possa se basear e aprender.

luizffgv replied:
Vou adicionar C++ nessa página. Acha que remover o UML completamente pode ser uma boa?

luizffgv committed:
Modificações para adequar à review

- Trocar termo "campo" para "atributo"

- Add tabela para esclarecer melhor a quais dados funções e métodos têm acesso

- Add exemplos e implementação em C e C++

- Trocar variável "c" por "cof"

luizffgv commented:
Depois muito tempo pensando eu acredito que minhas ideias para essa pull request não se adequariam ao esperado de um livro de C++, e tentar adaptá-las resultaria em um conteúdo de qualidade abaixo do ideal.

Gostaria de lembrar que meus commits foram feitos com a mesma licença (CC0), então se alguém futuramente quiser utilizar qualquer trecho que escrevi no meu fork sinta-se livre.
